[LV] Support strided load with a stride of -1

Mel-Chen opened the pull request:
Initial patch to support strided memory accesses. Currently, only stride -1 is supported, with future extensions planned for additional strides and strided store.

Mel-Chen committed:
Init: New Recipe VPWidenStridedLoadRecipe

alexey-bataev commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -1308,6 +1309,18 @@ class LoopVectorizationCostModel {
 
+  /// Returns true if the target machine can represent \p V as a strided load
+  /// or store operation.
+  bool isLegalStridedLoadStore(Value *V, ElementCount VF) {
```
I think better to have the format of the function, similar to isLegalGatherOrScatter, isLegalMaskedLoad etc.

Mel-Chen replied:
I copy the function name from TTI function. Do you have any recommended names?

Mel-Chen replied:
The prototype of isLegalGatherOrScatter is:
```
bool isLegalGatherOrScatter(Value *V, ElementCount VF)
```
And the prototype of isLegalMaskedLoad is:
```
bool isLegalMaskedLoad(Type *DataType, Value *Ptr, Align Alignment)
```
isLegalStridedLoadStore is the same as isLegalGatherOrScatter now.
Do you prefer the prototype of isLegalMaskedLoad?

alexey-bataev commented on llvm/lib/Transforms/Vectorize/VPlan.h:
```diff
@@ -2560,6 +2563,9 @@ class VPWidenMemoryRecipe : public VPRecipeBase {
   bool Reverse;
 
+  /// Whether the accessed addresses are evenly spaced apart by a fixed stride.
+  bool Strided;
```
```suggestion
  bool Strided = false;
```

alexey-bataev commented on llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:
```diff
@@ -2561,10 +2563,15 @@ InstructionCost VPWidenMemoryRecipe::computeCost(ElementCount VF,
+      return Ctx.TTI.getStridedMemoryOpCost(Ingredient.getOpcode(), Ty, Ptr,
+                                            IsMasked, Alignment, Ctx.CostKind,
+                                            &Ingredient);
+    else
```
```suggestion
```

Mel-Chen commented:
This patch is ready for review. Please take a look, thanks.

alexey-bataev commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -1312,6 +1313,22 @@ class LoopVectorizationCostModel {
 
+  /// Returns true if the target machine can represent \p V as a strided load
+  /// or store operation.
+  bool isLegalStridedLoadStore(Value *V, ElementCount VF) const {
```
I'm suggesting to use the prototype of isLegalMaskedLoad here, most of the function do the analysis based on type/alignment info

Mel-Chen replied:
After consideration, I have decided to rename `isLegalStridedLoadStore` to `stridedAccessCanBeWidened` and align its prototype with `memoryInstructionCanBeWidened` and `interleavedAccessCanBeWidened`. 
The reason is that, unlike `isLegalGatherOrScatter` and `isLegalMaskedLoad`, which primarily check whether the target supports certain operations, strided memory access requires additional checks, such as check stride. Its functionality is likely closer to `memoryInstructionCanBeWidened` and `interleavedAccessCanBeWidened`.

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:
```diff
@@ -2126,8 +2126,13 @@ void VPVectorPointerRecipe::execute(VPTransformState &State) {
+  // TODO: Support non-unit-reverse strided accesses.
+  Value *Index =
+      Strided
+          ? Builder.CreateMul(Increment, ConstantInt::getSigned(IndexTy, -1))
```
`Strided` means negative stride? Would be good to clarify in the comment for Strided..

alexey-bataev replied:
Just for now. Later it is supposed to support positive (and <-1 negative) and runtime stride values

lukel97 replied:
Should `Strided` eventually be some sort of `std::optional<int>`? Or is the plan to also allow runtime strided values? In which case I presume we need to model it as a VPValue operand.

Mel-Chen replied:
This is the initial patch for strided memory accesses, currently only supporting cases where the stride is -1. A follow-up patch will focus on stride analysis to support more strided memory accesses.

However, I believe it would be beneficial to discuss and decide how VPlan should handle stride in this initial patch. My current plan is to extend the existing `VPWidenLoadRecipe/VPWidenStoreRecipe` and `VPVectorPointerRecipe` by explicitly passing the stride into the recipe. For example:
```
VPWidenLoadRecipe(LoadInst &Load, VPValue *Addr, VPValue *Mask, 
                  VPValue *Stride,  
                  bool Consecutive, bool Reverse, DebugLoc DL)
```
A memory access will only be considered strided when `Stride` is explicitly provided (i.e., not nullptr).

@fhahn @ayalz What do you think about this approach?

fhahn replied:
I don't think we should add another 'mode' to WidenLoad/StoreRecipe to make it even more complicated.

It would probably be good to have a separate recipe.

lukel97 replied:
If we want to support it for EVL tail folding further down the line would we also need to add an EVL + Strided load recipe?

Mel-Chen replied:
> I don't think we should add another 'mode' to WidenLoad/StoreRecipe to make it even more complicated.
> 
> It would probably be good to have a separate recipe.

@fhahn I see. Let's first try adding VPStridedLoadRecipe for strided loads. What about VPVectorPointerRecipe? Can we modify it to:
```
VPVectorPointerRecipe(VPValue *Ptr, Type *IndexedTy, VPValue *Stride,  
                      GEPNoWrapFlags GEPFlags, DebugLoc DL) 
```
Or should we introduce a new recipe instead?

lukel97 replied:
@Mel-Chen yes, that's a good idea. Since it will generate a vp intrinsic either way.

fhahn replied:
It is a bit confusing that VPVectorPointerRecipe w/o stride steps by positive increment and with `Strided` by negative increment. Should it be VectorPointerEndRecipe?

Mel-Chen replied:
@fhahn Do you think we should introduce a new pointer recipe, or is it better to just extend the existing vector-pointer recipe? I prefer the later one.

fhahn replied:
I think it should be fine to add a stride argument for now, with a brief comment documenting the behavior.

wangpc-pp commented on llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:
```diff
@@ -2615,9 +2627,25 @@ void VPWidenLoadRecipe::execute(VPTransformState &State) {
+      Value *RuntimeVF =
+          getRuntimeVF(State.Builder, State.Builder.getInt32Ty(), State.VF);
+      NewLI = Builder.CreateIntrinsic(
+          Intrinsic::experimental_vp_strided_load, {DataTy, PtrTy, StrideTy},
```
I have a concern about how `experimental_vp_strided_load` is expanded when the target doesn't support it, and it is weird to use VP intrinsics in a non-VP code path.

Mel-Chen replied:
> I have a concern about how `experimental_vp_strided_load` is expanded when the target doesn't support it, and it is weird to use VP intrinsics in a non-VP code path.

This is because strided memory access does not have a non-VP intrinsic. However, we can set the %evl to the runtime VF, allowing it to benefit even when the tail is not folding by EVL.

lukel97 commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -6132,6 +6174,16 @@ void LoopVectorizationCostModel::setCostBasedWideningDecision(ElementCount VF) {
             ConsecutiveStride == 1 ? CM_Widen : CM_Widen_Reverse;
+        // Consider using strided load/store for consecutive reverse accesses to
+        // achieve more efficient memory operations.
+        if (ConsecutiveStride == -1 && stridedAccessCanBeWidened(&I, VF)) {
```
Is the `ConsecutiveStride == -1` redundant with the check already in `stridedAccessCanBeWidened`?

Mel-Chen replied:
I'm still deciding between `if (Legal->isConsecutivePtr(ScalarTy, Ptr) != -1)` and `if (!Legal->isConsecutivePtr(ScalarTy, Ptr))`.
Since, in terms of feasibility, both stride 1 and -1 can be used for strided memory access—it’s just a matter of potentially higher cost.

lukel97 commented on llvm/lib/Transforms/Vectorize/VPlan.h:
```diff
@@ -2642,16 +2652,16 @@ class VPWidenMemoryRecipe : public VPRecipeBase {
+                    bool Consecutive, bool Reverse, bool Strided, DebugLoc DL)
       : VPWidenMemoryRecipe(VPDef::VPWidenLoadSC, Load, {Addr}, Consecutive,
-                            Reverse, DL),
+                            Reverse, Strided, DL),
```
Do we need to update `onlyFirstLaneUsed` to return true for the address operand when Strided is true?

Mel-Chen replied:
Thanks for your reminder! Fixed it.
f77cc7e42c10211e2e279f3250c7eb19ae64d411

lukel97 commented on llvm/lib/Transforms/Vectorize/VPlan.h:
```diff
@@ -1571,12 +1571,15 @@ class VPVectorPointerRecipe : public VPRecipeWithIRFlags,
 public:
-  VPVectorPointerRecipe(VPValue *Ptr, Type *IndexedTy, GEPNoWrapFlags GEPFlags,
-                        DebugLoc DL)
+  VPVectorPointerRecipe(VPValue *Ptr, Type *IndexedTy, bool Strided,
```
Now that `VPVectorPointerRecipe` can compute the pointer for "reversed" strided loads/stores, should we clarify in the comments that `VPReverseVectorPointerRecipe` is for reversed strided loads/stores beginning at the last element?

Mel-Chen replied:
I agree that I should update the description of `VPVectorPointerRecipe`, but I'm not sure if `VPReverseVectorPointerRecipe` also needs to be modified. 
To me, `VPReverseVectorPointerRecipe` behaves like extracting the last pointer of each Part. 
If the comment for `VPReverseVectorPointerRecipe` needs to be updated, do you have any suggested new description?
217571003d0071186efdfe8c312c9c0cf2cd333c

lukel97 replied:
> To me, VPReverseVectorPointerRecipe behaves like extracting the last pointer of each Part for reverse accesses.

I think this is a good description, maybe it would be more clear if it was called something like `VPVectorEndPointerRecipe`? Not for this PR though, I might submit an NFC for this and see what people think

lukel97 replied:
I've opened https://github.com/llvm/llvm-project/pull/131086 for this

Mel-Chen commented:
Changes:
- Remove strided store functionality; will be supported in a future patch.
- Introduce VPWidenStridedLoadRecipe and explicitly pass the stride into the recipe.
Please take a look again, thanks.

lukel97 commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -8324,16 +8398,27 @@ VPRecipeBuilder::tryToWidenMemory(Instruction *I, ArrayRef<VPValue *> Operands,
   // reverse consecutive.
   LoopVectorizationCostModel::InstWidening Decision =
       CM.getWideningDecision(I, Range.Start);
+
+  auto SameWiden = [&](ElementCount VF) -> bool {
+    return Decision == CM.getWideningDecision(I, VF);
+  };
+  bool ContainsWidenVF =
+      LoopVectorizationPlanner::getDecisionAndClampRange(SameWiden, Range);
+  assert(ContainsWidenVF &&
+         "At least widen the memory accesses by the Start VF.");
+
```
Just to check my understanding, with this patch do we now have different VFs with different widening decisions that now need pruned? E.g. CM_Strided for VF=1, CM_WidenReverse for VF=2? Was this an issue beforehand?

Also is this assert still needed? I think this is only checking `CM.getWideningDecision(I, Range.Start) == CM.getWideningDecision(I, Range.Start)`

Mel-Chen replied:
> Just to check my understanding, with this patch do we now have different VFs with different widening decisions that now need pruned? E.g. CM_Strided for VF=1, CM_WidenReverse for VF=2? Was this an issue beforehand?

Yes, currently, when the VF is smaller, the cost model prefers CM_Strided, while for larger VFs, it favors CM_WidenReverse. This leads to a mismatch between the legacy cost model and the VPlan-based cost model, which is a situation we haven't encountered before.

> Also is this assert still needed? I think this is only checking `CM.getWideningDecision(I, Range.Start) == CM.getWideningDecision(I, Range.Start)`

I'm also torn about whether we should keep this assertion. ContainsWidenVF is theoretically supposed to be true; if it's not, we've definitely broken something. :(

lukel97 commented on llvm/lib/Transforms/Vectorize/VPlan.h:
```diff
@@ -2742,8 +2797,8 @@ struct VPWidenStoreRecipe final : public VPWidenMemoryRecipe {
            "Op must be an operand of the recipe");
-    // Widened, consecutive stores only demand the first lane of their address,
-    // unless the same operand is also stored.
+    // Widened, consecutive stores only demand the first lane of their
+    // address, unless the same operand is also stored.
```
Nit stray change

lukel97 reviewed (approved):
The new recipe LGTM

fhahn commented on llvm/test/Transforms/LoopVectorize/RISCV/riscv-vector-reverse.ll:
```diff
@@ -74,15 +74,15 @@ define void @vector_reverse_i64(ptr nocapture noundef writeonly %A, ptr nocaptur
+; CHECK-NEXT:      CLONE ir<[[ZEXT_IDX:%.+]]> = zext ir<[[IDX]]>
+; CHECK-NEXT:      CLONE ir<[[LD_IDX:%.+]]> = getelementptr inbounds ir<%B>, ir<[[ZEXT_IDX]]>
+; CHECK-NEXT:      vp<[[LD_PTR:%.+]]> = vector-pointer ir<[[LD_IDX]]>
+; CHECK-NEXT:      WIDEN ir<[[LD:%.+]]> = load vp<[[LD_PTR]]>, stride = ir<-4>, runtimeVF = vp<[[VF]]>
```
I am a bit surprised why this strides load with stride of 1 is more profitable than a consecutive load?

fhahn commented on llvm/lib/Transforms/Vectorize/VPlan.h:
```diff
@@ -2702,12 +2709,60 @@ struct VPWidenLoadEVLRecipe final : public VPWidenMemoryRecipe, public VPValue {
   }
 };
 
+/// A recipe for strided load operations, using the base address, stride, and an
+/// optional mask.
```
Conceptually the strided load with stride of 1 seems like it would be the same as a wide load? Would be good to clarify the difference.

alexey-bataev replied:
Possible in its current state, but not sure it is possible in general. Some (potential) strided memory recipes are currently represented as masked gathers, which require widened pointers. Strided/interleaved mem accesses do not require widened (vector) pointers, so it may affect the cost significantly. I think we need to teach the whole loop vectorizer the strided mem accesses pattern (no matter if the stride is constant, as it is right now, or runtime defined)

Mel-Chen replied:
Yes, you are right. Strided load can represent consecutive access, just like gather can also represent consecutive access. It's just that we usually don't choose the one with worse performance.
Updated the comment briefly. a2400038164625b4a4e888efc97b0584b57781e3

fhahn commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -1115,6 +1115,7 @@ class LoopVectorizationCostModel {
     CM_Widen_Reverse, // For consecutive accesses with stride -1.
     CM_Interleave,
     CM_GatherScatter,
+    CM_Strided,
```
Do we actually need all the legacy cost model changes? Could we just replace a `VPWidenLoadRecipe (vector-end-pointer) -> VPStridedLoad` if profitable as VPlan-to-VPlan transformation?

alexey-bataev replied:
Possible in its current state, but not sure it is possible in general. Some (potential) strided memory recipes are currently represented as masked gathers, which require widened pointers. Strided/interleaved mem accesses do not require widened (vector) pointers, so it may affect the cost significantly. I think we need to teach the whole loop vectorizer the strided mem accesses pattern (no matter if the stride is constant, as it is right now, or runtime defined)

Mel-Chen replied:
Besides the points @alexey-bataev mentioned, strided access might also need to be compared in terms of cost with interleaved access and gather/scatter access. If we encounter a situation where gather/scatter is converted to strided access, that's may workable, but converting interleaved groups to strided access might not be a simple task. Given all of this, I believe continuing with CM_Strided is a more scalable choice.

Mel-Chen replied:
I believe the main challenge is interleaved access and strided access.
The interleaved accesses: https://godbolt.org/z/eTcjvxYGK
In the same code, gather/scatter vector pointers with a stride of 2 can be transformed into strided loads/stores in the future: https://godbolt.org/z/hfhx953a8

If we realize during VPlan transforms that strided access is more profitable than interleaved access, converting interleaved accesses into strided ones would require splitting them back into multiple strided accesses.
Moreover, if an interleaved access originally requires a scalar epilogue due to a tail gap, 
```
void foo (int *A, int *B, int N) {
    int i;
    for (i = 0; i < N; ++i) {
        A[i*3] = B[i*3] + 1;
        A[i*3 + 1] = B[i*3 + 1] + 1;
        // No access [i*3+2], this is gap in the end of interleave group, requires scalar epilogue.
    }
}
```
such a requirement will disappear once it's converted into strided accesses. In that case, we would also need to modify the VPlan loop skeleton during the VPlan transform.
Given all that, the most we can do for now is something similar to what `VPTransform::createInterleaveGroup` currently does—decide whether to generate strided accesses before the VPlan is built, using an `enum InstWidening` to collect the necessary information, and then create a new `VPTransform::createStridedAccesses` to generate VPWidenStridedLoad/Store recipes.

Mel-Chen replied:
Ping @fhahn

fhahn replied:
Thanks for sharing the examples.

At least for interleave-groups, we already have a transform that removes interleave groups if possible (https://github.com/llvm/llvm-project/blob/main/llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp#L2685). I think the same could be done to replace it with strided/loads stores without too much complexity? Granted requiring scalar epilogue needs more thoughts, but that should also be solvable, and doesn't need to be included in the initial patches.

I've not yet checked the recipes generated for the gather-scatter example, but I think we should also be able to handle this in VPlan. If not, would be good to know at least what would be missing

lukel97 replied:
FWIW I've been experimenting locally with emitting strides as a VPlan transform. It's not a lot of code and it ends up more or less moving the stride detection from RISCVGatherScatterLowering.cpp into VPlan. It becomes easier though because instead of traversing through phis, you only really need to look for VPWidenGEPRecipes with a VPWidenIntOrFpInductionRecipe index instead (or VPScalarIVSteps). It looks something like this:

```c++
static void convertToStridedRecipes(VPlan &Plan) {
  SmallPtrSet<VPRecipeBase *, 4> ToDelete;
  LLVMContext &Ctx = Plan.getCanonicalIV()->getScalarType()->getContext();
  for (VPBasicBlock *VPBB : VPBlockUtils::blocksOnly<VPBasicBlock>(
           vp_depth_first_shallow(Plan.getVectorLoopRegion()->getEntry()))) {
    for (VPRecipeBase &R : make_early_inc_range(*VPBB)) {
      VPBuilder Builder(&R);

      if (auto *Store = dyn_cast<VPWidenStoreRecipe>(&R)) {
        if (Constant *StrideC = getStride(Store)) {
          VPValue *Stride = Plan.getOrAddLiveIn(StrideC);
          VPValue *Mask = Store->getMask();
          if (!Mask)
            Mask = Plan.getOrAddLiveIn(ConstantInt::getTrue(Ctx));
          VPValue *EVL =
              Builder.createScalarCast(Instruction::Trunc, &Plan.getVF(),
                                       Type::getInt32Ty(Ctx), R.getDebugLoc());
          auto *StridedStore = new VPWidenIntrinsicRecipe(
              Intrinsic::experimental_vp_strided_store,
              {Store->getStoredValue(), Store->getAddr(), Stride, Mask, EVL},
              Type::getVoidTy(Ctx));
          StridedStore->insertBefore(&R);
          R.eraseFromParent();
        }
      }

      if (auto *Load = dyn_cast<VPWidenLoadRecipe>(&R)) {
        if (Constant *StrideC = getStride(Load)) {
          VPValue *Stride = Plan.getOrAddLiveIn(StrideC);
          VPValue *Mask = Load->getMask();
          if (!Mask)
            Mask = Plan.getOrAddLiveIn(ConstantInt::getTrue(Ctx));
          VPValue *EVL =
              Builder.createScalarCast(Instruction::Trunc, &Plan.getVF(),
                                       Type::getInt32Ty(Ctx), R.getDebugLoc());
          auto *StridedLoad = new VPWidenIntrinsicRecipe(
              Intrinsic::experimental_vp_strided_load,
              {Load->getAddr(), Stride, Mask, EVL},
              Load->getIngredient().getType());
          StridedLoad->insertBefore(&R);
          R.getVPSingleValue()->replaceAllUsesWith(StridedLoad);
          R.eraseFromParent();
        }
      }
    }
  }
```

It doesn't necessarily need to use VPWidenIntrinsicRecipe, the strided recipe added in this PR would also work.

Mel-Chen replied:
> @fhahn Do we currently have a VPlan-based scalarization transformation? If not, do we need to install one before running `legalizeAndOptimizeInductions`?

Hmm...answered it by myself. Looks like just need some fix in `legalizeAndOptimizeInductions`. I will try to fix it tomorrow:)
```
    // Try to narrow wide and replicating recipes to uniform recipes, based on
    // VPlan analysis.
    // TODO: Apply to all recipes in the future, to replace legacy uniformity
    // analysis.
```

fhahn replied:
I recently put up https://github.com/llvm/llvm-project/pull/139150 to start in this direction. I have a few follow up patches that should get us close

Mel-Chen replied:
> I recently put up #139150 to start in this direction. I have a few follow up patches that should get us close

This is a good direction, but the current patch doesn't yet address the issue we're facing. 
So for now, I've made a small change in `legalizeAndOptimizeInductions`.a55afd43f8ef3d7536507fc2b5a23397e5b02365
While it's still not very effective at converting widen induction variables into derived ones in more complex use-chain scenarios, like:
```
; CHECK-NEXT:      ir<[[WIDEN_IV:%.+]]> = WIDEN-INDUCTION  ir<%n>, ir<-1>, vp<[[VF]]>
; CHECK-NEXT:      WIDEN ir<[[IDX:%.+]]> = add nsw ir<[[WIDEN_IV]]>, ir<-1>
; CHECK-NEXT:      WIDEN-CAST ir<[[ZEXT_IDX:%.+]]> = zext ir<[[IDX]]> to i64
; CHECK-NEXT:      CLONE ir<[[LD_IDX:%.+]]> = getelementptr inbounds ir<%B>, ir<[[ZEXT_IDX]]>
; CHECK-NEXT:      vp<[[LD_PTR:%.+]]> = vector-pointer ir<[[LD_IDX]]>
; CHECK-NEXT:      WIDEN ir<[[LD:%.+]]> = load vp<[[LD_PTR]]>, stride = ir<-4>, runtimeVF = vp<[[VF]]>
```
, it at least allows us to continue to try this path and ensures that the lit test cases of EVL folding remain vectorizable.

Mel-Chen commented on llvm/lib/Transforms/Vectorize/VPlan.h:
```diff
@@ -2702,12 +2709,60 @@ struct VPWidenLoadEVLRecipe final : public VPWidenMemoryRecipe, public VPValue {
-    // Widened loads only demand the first lane of EVL and consecutive loads
-    // only demand the first lane of their address.
+    // Widened loads only demand the first lane of EVL and consecutive/strided
+    // loads only demand the first lane of their address.
```
Dropped them, thanks!

Mel-Chen commented:
Ping @fhahn @ayalz

Mel-Chen committed:
[WIP][VPlan Based] Try to remove CM_Strided from uniform analysis

Also cherry-pick the branch Mel-Chen:legalizeAndOptimizeInductions
However, still not work well as collectLoopUniforms if the use-chain is
too compilicated. :(

Mel-Chen committed:
[WIP][VPlan Based] Generate VPWidenStrideLoadRecipe in VPlanTransform

Still rely on CM_Strided to known legal and cost.

Mel-Chen committed:
[WIP][VPlan based] Clamp VF range in VPlan transformation

Mel-Chen committed:
[WIP][VPlan based] Time to remove CM_Strided

Mel-Chen committed:
[VPlan based] Patch comments, nfc

Mel-Chen committed:
Format

Mel-Chen commented:
I’ve removed CM_Strided and moved the generation of strided access recipes entirely into the VPlan transformation phase. Please take a look, thanks.

Here are some potential issues that are currently masked because we only convert reverse accesses at the moment, but they may need to be addressed in the future:
1. Uniform and scalar analysis:
If the original memory recipe is not consecutive, the address for strided accesses is still treated as a vector address instead of a scalar one. This will result in suboptimal performance.

2. Interleaved accesses:
At the moment, interleave groups are determined by the legacy cost model, and they are created after the conversion to strided accesses. However, some interleaved accesses could be converted into multiple strided accesses instead. To support this, we’ll need to revisit the order and interaction between generating interleave groups and strided recipes.

Mel-Chen commented:
Ping

fhahn commented on llvm/lib/Transforms/Vectorize/VPlan.h:
```diff
@@ -1720,16 +1723,21 @@ class VPVectorEndPointerRecipe : public VPRecipeWithIRFlags,
 };
 
 /// A recipe to compute the pointers for widened memory accesses of IndexTy.
+/// Supports both consecutive and reverse consecutive accesses.
```
Is this still accurate? VPVectorPointerEndRecipe also is for reversed accesses?

fhahn commented on llvm/lib/Transforms/Vectorize/VPlan.h:
```diff
@@ -1720,16 +1723,21 @@ class VPVectorEndPointerRecipe : public VPRecipeWithIRFlags,
                               public VPUnrollPartAccessor<1> {
   Type *IndexedTy;
 
+  /// Indicate whether to compute the pointer for strided memory accesses.
```
Should document what the stride is.

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:
```diff
@@ -3072,6 +3087,48 @@ void VPWidenLoadEVLRecipe::print(raw_ostream &O, const Twine &Indent,
+  const Align Alignment = getLoadStoreAlignment(&Ingredient);
+
+  auto &Builder = State.Builder;
+  State.setDebugLocFrom(getDebugLoc());
```
Shouldn't be needed, it is set before each recipe is executed

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -2519,6 +2530,75 @@ void VPlanTransforms::dissolveLoopRegions(VPlan &Plan) {
+        const InstructionCost CurrentCost = MemR->computeCost(VF, Ctx);
+        const InstructionCost StridedLoadStoreCost =
+            Ctx.TTI.getStridedMemoryOpCost(
+                Ingredient.getOpcode(), DataTy,
```
can we get the opcode from the recipe instead the ingredient?

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -2519,6 +2530,75 @@ void VPlanTransforms::dissolveLoopRegions(VPlan &Plan) {
+        const InstructionCost StridedLoadStoreCost =
+            Ctx.TTI.getStridedMemoryOpCost(
+                Ingredient.getOpcode(), DataTy,
+                getLoadStorePointerOperand(&Ingredient), MemR->isMasked(),
```
Would be good to avoid passing the pointer operand from the ingredient here, as it may not be accurate any longer. What does the interface use the pointer operand for? Ideally we would pass the required info directly w/o having to pass an IR value. Do you think that would be feasible?

Otherwise, can we pass the underlying value from the pointer operand, if it exists?

Mel-Chen replied:
`getStridedMemoryOpCost` falls back to `getGatherScatterOpCost` when the target doesn't support strided access.
```
  InstructionCost getStridedMemoryOpCost(unsigned Opcode, Type *DataTy,
                                         const Value *Ptr, bool VariableMask,
                                         Align Alignment,
                                         TTI::TargetCostKind CostKind,
                                         const Instruction *I) const override {
    // For a target without strided memory operations (or for an illegal
    // operation type on one which does), assume we lower to a gather/scatter
    // operation.  (Which may in turn be scalarized.)
    return thisT()->getGatherScatterOpCost(Opcode, DataTy, Ptr, VariableMask,
                                           Alignment, CostKind, I);
  }
```
Since `getGatherScatterOpCost` uses pointer operand, I went ahead with the latter approach you suggested.
8875c321ac05f6700b41baa377bdcca8efd45caf

Mel-Chen committed:
[Unrelated code] Remove Mel-Chen:legalizeAndOptimizeInductions

We should reopen it after supporting const strided accessess.

Mel-Chen committed:
[Fix] Remove unused debug info setting.

Mel-Chen committed:
[Fix] Set Opcode as Instruction::Load directly.

Mel-Chen committed:
[Fix] Replace getLoadStorePointerOperand with Ptr->getUnderlyingValue().

Mel-Chen committed:
[Fix] Pass stride in element size

Mel-Chen committed:
[Fix] New operand Stride for VPVectorPointerRecipe

fhahn commented on llvm/lib/Transforms/Vectorize/VPlan.h:
```diff
@@ -1743,18 +1746,20 @@ class VPVectorEndPointerRecipe : public VPRecipeWithIRFlags,
 
 /// A recipe to compute the pointers for widened memory accesses of IndexTy.
```
Would be good to document the arguments

Mel-Chen replied:
b7828450869547398a4e2cb0c2b926dce6d479e7

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -632,7 +632,6 @@ static void legalizeAndOptimizeInductions(VPlan &Plan) {
     auto *PhiR = dyn_cast<VPWidenInductionRecipe>(&Phi);
     if (!PhiR)
       continue;
-
```
unrelated change

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -2523,6 +2529,78 @@ void VPlanTransforms::dissolveLoopRegions(VPlan &Plan) {
+
+      // The stride of consecutive reverse access must be -1.
+      int64_t Stride = -1;
+      auto *GEP = dyn_cast<GetElementPtrInst>(PtrUV->stripPointerCasts());
```
Is stripping the pointer casts needed here? If it is just needed for the flags, can we get the wrap flags from the recipe?

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -2523,6 +2529,78 @@ void VPlanTransforms::dissolveLoopRegions(VPlan &Plan) {
+      auto *MemR = dyn_cast<VPWidenMemoryRecipe>(&R);
+      // TODO: support strided store
+      // TODO: support strided accesses with stride not equal to -1
+      if (!MemR || !isa<VPWidenLoadRecipe>(MemR) || !MemR->isReverse())
```
Do we need to check if MemR is consecutive here?

Mel-Chen replied:
I think we don't have to check if MemR is consecutive. There is an assertion in constructor of VPWidenMemoryRecipe:
```
assert((Consecutive || !Reverse) && "Reverse implies consecutive");
```
So I add an assertion for consecutive.
72c17a195f0d58be93ebd405a6f09857102cf704

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -2523,6 +2529,78 @@ void VPlanTransforms::dissolveLoopRegions(VPlan &Plan) {
+
+      auto *VecEndPtr = cast<VPVectorEndPointerRecipe>(MemR->getAddr());
+      VPValue *Ptr = VecEndPtr->getPtr();
+      Value *PtrUV = Ptr->getUnderlyingValue();
```
Ah I see, the Ptr argument isn't used by the RISCV TTI, but is potentially used by others; it appears passing nullptr here would work, but then it may cause issues with TTI implementations that use the default from BasicTTI?

Mel-Chen replied:
Yes, RISC-V TTI can omit passing the pointer because it implements getStridedMemoryOpCost and does not require full pointer information, so passing nullptr could work. However, this might cause issues for targets that do not implement getStridedMemoryOpCost.

Mel-Chen committed:
[Comment] Remove the unrelated change, nfc

Mel-Chen committed:
[Comment] Add assert for consecutive, nfc

Mel-Chen committed:
[Comment] Update comment of VPVectorPointerRecipe

Mel-Chen commented:
ping

fhahn commented:
Do you know if/how this is related/interacts with https://github.com/llvm/llvm-project/pull/130032?

LiqinWeng commented:
> Do you know if/how this is related/interacts with #130032?

Sorry， Sorry, I'll add a description.

Mel-Chen commented:
ping
